opt: Push filter down into the set operators

ridwanmsharif opened the pull request:
This commit adds a  norm rule that pushes down filters below the
set operators as they are logically equivalent but will yield
better performance. The set operators that will push down the filters
are: Union, Union All, Except, Except All, Intersect, Intersect All.

See the comment on the PushFilterIntoSetOp for more detailed
justification for the rule.

Release note: None

Commit:
opt: Push filter down into the set operators

This commit adds a norm rule that pushes down filters below the
set operators as they are logically equivalent but will yield
better performance. The set operators that will push down the filters
are: Union, Union All, Except, Except All, Intersect, Intersect All.

See the comment on the PushFilterIntoSetOp for more detailed
justification for the rule.

Release note: None

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37418) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37418) before we can accept your contribution.<br/><hr/>**Ridwan Sharif** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=37418) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37418)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @justinj and @ridwanmsharif)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LePldHp6adQA2jBehYR:-LePldHp6adQA2jBehYS:b3gpace) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
> ```Go
> 			}
> 			return c.f.ConstructVariable(opt.ColumnID(dstCol))
> 		case *memo.SubqueryExpr, *memo.ExistsExpr, *memo.AnyExpr:
> ```

If these subquery operators are correlated (i.e. have outer variable references), they need to have variable references replaced. Here's an example:
```
SELECT *
FROM (
  SELECT * FROM t
  UNION ALL
  SELECT * FROM t
) t2
WHERE
  CASE
    WHEN t2.b = 0
    THEN EXISTS(SELECT * FROM t WHERE a=t2.a / t2.b)
    ELSE null
  END;
```
Here, the CASE statement contains a possible division by 0 error, and so can't be decorrelated according to our side effect policy (see logical.go, `CanHaveSideEffects` comment).

I think the fix is to just remove those special cases. But these things can be tricky, so definitely needs some experimentation and tests (I'd add a test for each of those subquery operators).
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1081 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LePmif9AweuBDENRi-m:-LePmif9AweuBDENRi-n:bqz6da5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1081)):*
> ```Go
> 		return c.f.Replace(nd, replace)
> 	}
> 	// Replace every FilterItem in the filters expression before creating new FilterExpr
> ```

NIT: In opt code, we typically add a blank line before a comment that is not the first line in a block.


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
This looks great! You got that done really fast. Mostly some comment nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeRdAsMDl7qC4KtCT3K:b-6g3pf3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @ridwanmsharif)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1049 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRTvWF9iy93pGh-jLH:-LeRTvWF9iy93pGh-jLI:b-87hfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1049)):*
> ```Go
> //   SELECT * FROM (SELECT x FROM a UNION SELECT y FROM b) WHERE x < 5
> //
> // If setMap is called on the left subtree of the Union, the filter x < 5
> ```

This claim seems questionable, or at least misleading, to me. The fact that the *name* `x` is re-used isn't a property of the opt tree, but of the original SQL query, and is not relevant any more at this point (the fact that sometimes the outcol list is the same as the left col list is an optimization to save column IDs and is unrelated to the naming rules of SQL). What's relevant now is that we translate `VariableExpr`s based on the set's private, which may or may not cause a rewrite of the column IDs (and I thought in the case of Union specifically we always create new column IDs).
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1065 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRcgI47vzFWADKf4yG:-LeRcgI47vzFWADKf4yH:b9cnqza) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1065)):*
> ```Go
> 	// that need to be replaced and then replace them appropriately.
> 	var replace ReplaceFunc
> 	replace = func(nd opt.Expr) opt.Expr {
> ```

I wonder if this whole thing could be shared somehow between this and the join version? Might not be a way to make that work, though
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1070 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRcTqB4kt-fdMjT-Iu:-LeRcTqC8d88TZBUud18:b-47wca9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1070)):*
> ```Go
> 			dstCol, inCol := colMap.Get(int(t.Col))
> 			if !inCol {
> 				// Avoid constructing a new variable if its not part of the out cols.
> ```

nit: `it's`.

This comment (particularly "Avoid constructing") makes this read like this is an optimization to me, but in reality it's just that there's nothing to do here, right? It's just not a column that we're concerned with.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LePldHp6adQA2jBehYR:-LeRRa8e7Slxgmwhe-pu:b-h58jua) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If these subquery operators are correlated (i.e. have outer variable references), they need to have variable references replaced. Here's an example:
```
SELECT *
FROM (
  SELECT * FROM t
  UNION ALL
  SELECT * FROM t
) t2
WHERE
  CASE
    WHEN t2.b = 0
    THEN EXISTS(SELECT * FROM t WHERE a=t2.a / t2.b)
    ELSE null
  END;
```
Here, the CASE statement contains a possible division by 0 error, and so can't be decorrelated according to our side effect policy (see logical.go, `CanHaveSideEffects` comment).

I think the fix is to just remove those special cases. But these things can be tricky, so definitely needs some experimentation and tests (I'd add a test for each of those subquery operators).
</blockquote></details>

Ah right, I told Ridwan not to worry about correlated subqueries but I had been thinking of cases where we were inside of a correlated subquery higher up.
___
*[pkg/sql/opt/norm/rules/set.opt, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRc99A7hD7SQ3GJKRs:-LeRc99A7hD7SQ3GJKRt:bxpjrxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L36)):*
> ```
> 
> # PushFilterIntoSetOp pushes a filter into both the left and right sides
> # of all set operators after being mapped to use the columns of both sides.
> ```

I think conceptually we're just pushing the filter down and that's what this comment should say. That we have to rename some variables for that to be valid is an implementation detail.
___
*[pkg/sql/opt/norm/rules/set.opt, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRZ9FH9H1msW5b5MAh:-LeRZ9FI7NZjfLW63vs7:b-u6121d) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L44)):*
> ```
> # in the union the x out col corresponds to the x column in the (a) table and
> # the y column in the (b) table. Therefore our mapping becomes x < 5 in the
> # left side, and y <  5 in the right side.
> ```

nit: extra space.

I think you can compress this and still have it be understandable—if you just show an example of the transformation the reader can understand the mapping that's occurred:

```
This rule performs the following transformation:

  SELECT * FROM (SELECT x FROM a UNION ALL SELECT y FROM b) WHERE x < 5
=>
  (SELECT x FROM a WHERE x < 5) UNION ALL (SELECT y FROM b WHERE y < 5)
```
___
*[pkg/sql/opt/norm/rules/set.opt, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRZd5Z8HadeCPb33bx:-LeRZd5Z8HadeCPb33by:bj2tqcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L52)):*
> ```
> # this claim:
> # 
> # Two relations are union compatible when they have the same number of
> ```

I'm not sure this justification is necessary. If you want to explain why this works you could just say like, "because set operators don't modify rows, expressions valid for the result can also be applied to the inputs," but I think most readers of this code will understand that this is how the set operators work.
___
*[pkg/sql/opt/norm/rules/set.opt, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRXp6WDcr_MesF_qDu:-LeRXp6WDcr_MesF_qDv:b8d7ne8) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L67)):*
> ```
> # For some row to be in the result set, it must satisfy the filter condition
> # and the condition logically implied by the set operators (Union, Except,
> # Intersect). These conditions are independent and so the order in which they
> ```

This is cool, but I'm not sure this amount of detail is necessary - I would just say something like:

```
This works in the case of Union because a row passing the filter is independent of the other rows in its relation. It works in the case of Intersect because if a row `a` was filtered from the intersection, it will also be filtered from the LHS and RHS. It works in the case of Except because if a row is filtered from the RHS, preventing its removal from the LHS, it will also have been filtered from the LHS.  
```
___
*[pkg/sql/opt/norm/testdata/rules/set, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRe-ee6yEILNb6WY0i:-LeRe-ee6yEILNb6WY0j:bbvoqy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/testdata/rules/set#L13)):*
> ```
>       └── k int not null
> 
> exec-ddl
> ```

I would also make sure that there are some logic tests that exercise this code. See `pkg/sql/logictest/testdata/logic_test/union`. You can run them with `make testlogic FILES=union`.
___
*[pkg/sql/opt/norm/testdata/rules/set, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRhER31B12qy6steRX:-LeRhER41xyOikmNQ125:b4n59q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/testdata/rules/set#L74)):*
> ```
> 
> # --------------------------------------------------
> # PushFilterIntoSetOp
> ```

I think some tests that remap more than one (or zero) variables would also be good!
___
*[pkg/sql/opt/norm/testdata/rules/set, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRbKyK1i50A36YUGB3:-LeRbKyK1i50A36YUGB4:b-ih9r48) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/testdata/rules/set#L237)):*
> ```
> 
> opt expect=PushFilterIntoSetOp
> SELECT k FROM
> ```

Consider adding another test case involving multiple set ops, where the filter has to travel down further than one level


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Wow, great work already!

There are some annoying cornercases to consider:
 - some types can have equal but non-identical values, filters on those can't be remapped in general. Here's an example:
`select * from ((values (1.0::decimal)) except (values (1.00::decimal))) where column1::string != '1.00';`
This would return 0 results without pushing down, but 1 result with push down. The join filter remap rules check this property (search for `DatumTypeHasCompositeKeyEncoding`call in `GetEquivColsWithEquivType`)

- if there are side-effects in the filter, we don't want to duplicate it. Example:
`select * from ((values (1,2)) except (values (0,1))) where 1 / column1 > 0;`
 Maybe it's ok for `union all`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeS2P00EXIK_0mQsreT:b-jynv3k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @ridwanmsharif)



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
Our side-effect policy allows duplication (see `CanHaveSideEffects` comment in `logical.go`). If a user needs to control the number of times a particular expression will be evaluated, the idea is that they would use a CTE (though I recognize we don't implement those correctly yet). It will be whack-a-mole (and will tie our hands) if we try to guarantee the number of times an expression will be evaluated, so we want to put that burden on the user. See `PushFilterIntoJoinLeftAndRight` for another example where we duplicate a filter.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeS9yiN1Kcgj4-aCPqm:b-1j5lgg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @ridwanmsharif)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Indeed, forget my second point. Justin and I found a case where it does this.

```
radu=> select * from a; 
a 
---
 1
 2
(2 rows)

radu=> select * from b;
 b 
---
 0
 1
(2 rows)

radu=> select * from ((select * from a) intersect (select * from b)) AS foo where 1 / a > 0;
ERROR:  division by zero
radu=> select * from ((select * from a) intersect (select * from b)) AS foo;
 a 
---
 1
(1 row)
```

Commit:
opt: don't push unmappable filters into set ops

This commit adds onto the norm rule for pushing filters
into set operators. Specifically, it makes sure to only pass
down filters that don't have composite types. The remaining
filters are not passed down.

Release note: None

ridwanmsharif reviewed (commented):
It still looks like it makes more sense to not pass down filters that have side effects. The check for it is trivial and is does seem more correct even if postgres does push it down. Is there a reason to not push filters down other than similarity to postgres?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeUp5j0AWWdA09plJ_3:b-7pnt68)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @justinj)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1049 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRTvWF9iy93pGh-jLH:-LeUiAB07mNrgMS158mC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1049)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This claim seems questionable, or at least misleading, to me. The fact that the *name* `x` is re-used isn't a property of the opt tree, but of the original SQL query, and is not relevant any more at this point (the fact that sometimes the outcol list is the same as the left col list is an optimization to save column IDs and is unrelated to the naming rules of SQL). What's relevant now is that we translate `VariableExpr`s based on the set's private, which may or may not cause a rewrite of the column IDs (and I thought in the case of Union specifically we always create new column IDs).
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1065 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRcgI47vzFWADKf4yG:-LeUhEXlEgzFijqYGv9F:b3wynow) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1065)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I wonder if this whole thing could be shared somehow between this and the join version? Might not be a way to make that work, though
</blockquote></details>

Probably best to keep it as it, they differ when it comes to sub expressions.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LePldHp6adQA2jBehYR:-LeUhPUX-lC8yJ2zrHjw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1074)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Ah right, I told Ridwan not to worry about correlated subqueries but I had been thinking of cases where we were inside of a correlated subquery higher up.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/set.opt, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRc99A7hD7SQ3GJKRs:-LeUk3ee6jO1ISM8AKrS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L36)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think conceptually we're just pushing the filter down and that's what this comment should say. That we have to rename some variables for that to be valid is an implementation detail.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/set.opt, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRZd5Z8HadeCPb33bx:-LeUjdQe79FXXZZ-Q0kG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L52)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I'm not sure this justification is necessary. If you want to explain why this works you could just say like, "because set operators don't modify rows, expressions valid for the result can also be applied to the inputs," but I think most readers of this code will understand that this is how the set operators work.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/set.opt, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRXp6WDcr_MesF_qDu:-LeUk2S1722Au_r9Hu6A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/rules/set.opt#L67)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This is cool, but I'm not sure this amount of detail is necessary - I would just say something like:

```
This works in the case of Union because a row passing the filter is independent of the other rows in its relation. It works in the case of Intersect because if a row `a` was filtered from the intersection, it will also be filtered from the LHS and RHS. It works in the case of Except because if a row is filtered from the RHS, preventing its removal from the LHS, it will also have been filtered from the LHS.  
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/testdata/rules/set, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRe-ee6yEILNb6WY0i:-LeUhUGfDPwmWwPzj9cX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/testdata/rules/set#L13)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I would also make sure that there are some logic tests that exercise this code. See `pkg/sql/logictest/testdata/logic_test/union`. You can run them with `make testlogic FILES=union`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/testdata/rules/set, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRhER31B12qy6steRX:-LeUhVAm8IjAfEasbwlm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/testdata/rules/set#L74)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think some tests that remap more than one (or zero) variables would also be good!
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/testdata/rules/set, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRbKyK1i50A36YUGB3:-LeUhW6DDvMEnwQJvUTg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/testdata/rules/set#L237)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Consider adding another test case involving multiple set ops, where the filter has to travel down further than one level
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
It's simple in this case, but in many others it might not be. We just don't want our code to become a mishmash between guaranteeing something and not guaranteeing something. Imagine someone who looks at the code for the first time and finds this mismatch between joins and set operations. It's better to stick to the same principle everywhere.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeWczGg5UeiEYcJheCq:bi13vky)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, and @ridwanmsharif)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1062 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeWerJ61uhfh2MQO8fN:-LeWerJ61uhfh2MQO8fO:bjppb5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d8831bd50261fa2550721436a59b641009c495d/pkg/sql/opt/norm/custom_funcs.go#L1062)):*
> ```Go
> // HasMappableFilters determines whether at least one filter in the FiltersExpr
> // can be mapped to either side of a set operator.
> func (c *CustomFuncs) HasMappableFilters(filters memo.FiltersExpr) bool {
> ```

As a general rule, we try to make the "match" part of the rule as efficient as possible (that part runs every time). In particular, we want to avoid allocating stuff just to check if a rule matches. This might mean we need to duplicate a bit of code (in this case it would be part of the loop with an early return). Or separatePushDownFilters could be improved to not allocate in the case where nothing is pushable and just return `nil, filters`.

I realize that for this particular rule we expect the common case to be a match, so maybe it's not important here.
___
*[pkg/sql/opt/norm/rules/set.opt, line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeWeYkS7_D6w-gt0CPL:-LeWeYkS7_D6w-gt0CPM:b2rzg7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d8831bd50261fa2550721436a59b641009c495d/pkg/sql/opt/norm/rules/set.opt#L74)):*
> ```
> # Intersection (All): (A Intersect B) \ C => (A \ C) Intersect (B \ C)
> # Except (All):       (A Except B) \ C => (A \ C) Except (B \ C)
> [PushFilterIntoSetOp, Normalize]
> ```

I'm wondering why this doesn't look more like `PushFilterIntoJoinLeftAndRight`. It seems cleaner to rely on optgen to find an item that can be pushed down, instead of working with the entire filters list ourselves.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
[nit] in the commit message and PR description: theya re -> they are

In general I agree with Radu and Andy's comments that you can take a look at `PushFilterIntoJoinLeftAndRight` for some ideas here. In particular, take a look at how the `CanMap` function works to see how it does as little as possible to determine whether or not the rule should be applied. You can also follow that model of pushing down one filter at a time, rather than trying to do all filters at once.

I also agree that we should be consistent with the join filter push-down rules and not worry about side effects.

> Reviewed 2 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LekltUs8hPQDxJPTK4Q:batasih)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @ridwanmsharif)



<!-- Sent from Reviewable.io -->

Commit:
opt: push filters down one at a time into set op

This change makes it so that filters are passed one at a time instead
of all together into set operators. This is for more consistency with
the Join filters. Since the filter list is expected to be small, we're
not too worried about the extra intermediate steps.

Release note: None

ridwanmsharif reviewed (commented):
Filters are now passed down one at a time since we're dealing with relatively small number of filter items. Its cleaner this way and similar to `PushFilterIntoLeftAndRight`. 
I'm a little concerned about the fact that this PR makes it so that the optimizer spits out a non-logically equivalent evaluation of an expression compared to if the optimizer was turned off when side-effects are concerned. 
Like in Radu's example before `select * from ((values (1,2)) except (values (0,1))) where 1 / column1 > 0;` would return (1, 2) before the change but would fail (with div by 0) after the change. It does seem like that's okay as postgres does this too so I defer to the teams judgement.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeXGVrfDmn2mVB1co6B:b-u8j275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/norm/rules/set.opt, line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeWeYkS7_D6w-gt0CPL:-LemZFJbDNq8xSZMJRJT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d8831bd50261fa2550721436a59b641009c495d/pkg/sql/opt/norm/rules/set.opt#L74)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm wondering why this doesn't look more like `PushFilterIntoJoinLeftAndRight`. It seems cleaner to rely on optgen to find an item that can be pushed down, instead of working with the entire filters list ourselves.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice changes!

[nit] in the second commit message: so gthaty filters -> so that filters
Also fix theya re in the first commit message (looks like you already fixed it in the PR message)

I think it's best to stick with the Postgres behavior, especially since it's consistent with our side effect policy. It might be good to actually add Radu's example as a test case if you haven't already just so we are explicit about this behavior.

> Reviewed 4 of 6 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-Len6DK_7ZdEog0odvS9:b-ar2mmq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @ridwanmsharif)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1025 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Len8egt8w9iQWHC62ws:-Len8egt8w9iQWHC62wt:b-k0rvam) ([raw file](https://github.com/cockroachdb/cockroach/blob/94930917fa879cc682713e27ce888eaab3116407/pkg/sql/opt/norm/custom_funcs.go#L1025)):*
> ```Go
> 	filterProps := src.ScalarProps(c.mem)
> 	hasCompositeKeyEncoding := false
> 	filterProps.OuterCols.ForEach(func(i int) {
> ```

Instead of `ForEach`, you can use this pattern to save some cycles:
```
for i, ok := filterProps.OuterCols.Next(0); ok; i, ok = filterProps.OuterCols.Next(i + 1) {
	colType := c.f.Metadata().ColumnMeta(opt.ColumnID(i)).Type
	if sqlbase.DatumTypeHasCompositeKeyEncoding(colType) {
		return false
	}
}
```
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1037 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LenAFyVAtiWOZvPlPJc:-LenAFyVAtiWOZvPlPJd:b-c1k8su) ([raw file](https://github.com/cockroachdb/cockroach/blob/94930917fa879cc682713e27ce888eaab3116407/pkg/sql/opt/norm/custom_funcs.go#L1037)):*
> ```Go
> }
> 
> // MapFilterOnLeft replaces the columns in the filter that are in the OutCols
> ```

[nit] replaces -> maps, in -> on
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1046 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LenAT9G7WAkoa-syFgN:-LenAT9G7WAkoa-syFgO:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/94930917fa879cc682713e27ce888eaab3116407/pkg/sql/opt/norm/custom_funcs.go#L1046)):*
> ```Go
> }
> 
> // MapFilterOnRight replaces the columns in the filter that are in the OutCols
> ```

ditto


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
Fixed the commits now

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LeoJh5X0DQyckOtvuQv:b-s4lez2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1025 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Len8egt8w9iQWHC62ws:-LeoJw1vBvJRJYAKdeOC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94930917fa879cc682713e27ce888eaab3116407/pkg/sql/opt/norm/custom_funcs.go#L1025)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Instead of `ForEach`, you can use this pattern to save some cycles:
```
for i, ok := filterProps.OuterCols.Next(0); ok; i, ok = filterProps.OuterCols.Next(i + 1) {
	colType := c.f.Metadata().ColumnMeta(opt.ColumnID(i)).Type
	if sqlbase.DatumTypeHasCompositeKeyEncoding(colType) {
		return false
	}
}
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1037 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LenAFyVAtiWOZvPlPJc:-LeoJwwU8hzTKzKo3TL1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94930917fa879cc682713e27ce888eaab3116407/pkg/sql/opt/norm/custom_funcs.go#L1037)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] replaces -> maps, in -> on
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1046 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LenAT9G7WAkoa-syFgN:-LeoJyruAjWhREbJoN5r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94930917fa879cc682713e27ce888eaab3116407/pkg/sql/opt/norm/custom_funcs.go#L1046)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LerBEfu-7fDg5omd-6m:bo31fz2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1065 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRcgI47vzFWADKf4yG:-LerA4288WE_P01B2GMq:btrc8e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1065)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Probably best to keep it as it, they differ when it comes to sub expressions.
</blockquote></details>

Both have this common piece of code that remaps columns using a FastIntMap. What is different? the `case *memo.SubqueryExpr, *memo.ExistsExpr, *memo.AnyExpr`? Can't we do that here too, now that we don't remap when there are correlated subqueries? If we still need it, we can just have a `skipSubqueries` flag in the common function.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1022 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Ler9dWh5pOm3WaFWH--:-Ler9dWi-GuKs58tWBat:b-6if3te) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1022)):*
> ```Go
> // CanMapOnSetOp determines whether the filter can be mapped to either
> // side of a set operator.
> func (c *CustomFuncs) CanMapOnSetOp(src *memo.FiltersItem) bool {
> ```

`CanMapSetOpFilter`

By the way, I'd also consider renaming the general `CanMap` and `Map` calls to something that includes "join"
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1037 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Ler9fqsBTX6VwOdm47z:-Ler9fqsBTX6VwOdm48-:bfuhvwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1037)):*
> ```Go
> // left side of the operator.
> // Useful for pushing filters to relations the set operation is composed of.
> func (c *CustomFuncs) MapFilterOnLeft(
> ```

[nit] `MapSetOpFilterLeft`
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1056 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LerBRYJ59N6cJocb_7Z:-LerBRYJ59N6cJocb_7_:b7vbkne) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1056)):*
> ```Go
> // expression dst instead of the columns in src.
> //
> // setMap assumes the two relations are union compatible.
> ```

This description doesn't quite fit with the function. The function doesn't even know of any relations, it just rebuilds a filter with remapped columns.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1069 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Ler9w2_22_xTG99_Zgl:-Ler9w2_22_xTG99_Zgm:b-qz10f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1069)):*
> ```Go
> // If setMap is called on the right subtree, the filter x < 5 will be mapped
> // similarly to y < 5 on the right side.
> func (c *CustomFuncs) setMap(
> ```

[int] `mapSetOpFilter` ore `remapFilter`


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 6 files at r8, 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LestMSh14mN2Gh21K8f:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @ridwanmsharif)



<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LetNiCCDctZyS8l6Azl:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1065 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LeRcgI47vzFWADKf4yG:-LetNt9t5SeHfi1dvKMC:b-1pyp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd95a3bc2606c3b1beb08ef9b7e8d3bb99412f95/pkg/sql/opt/norm/custom_funcs.go#L1065)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Both have this common piece of code that remaps columns using a FastIntMap. What is different? the `case *memo.SubqueryExpr, *memo.ExistsExpr, *memo.AnyExpr`? Can't we do that here too, now that we don't remap when there are correlated subqueries? If we still need it, we can just have a `skipSubqueries` flag in the common function.
</blockquote></details>

Its this difference and also the colSets they use and how they use them. Its also a standard way of walking the tree as described in the comment over `Replace`. I don't know if it is general enough and worth pulling out.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1037 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Ler9fqsBTX6VwOdm47z:-LetNo2N4AJcJUSBWq8D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1037)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `MapSetOpFilterLeft`
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1056 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LerBRYJ59N6cJocb_7Z:-LetNpatAlVMFQVVfUW2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1056)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This description doesn't quite fit with the function. The function doesn't even know of any relations, it just rebuilds a filter with remapped columns.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1069 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-Ler9w2_22_xTG99_Zgl:-LetNqq64qOLsMKOwR5W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1069)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[int] `mapSetOpFilter` ore `remapFilter`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-:-LetUQzk9OYz071kPdLA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @justinj, @ridwanmsharif, and @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1056 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LerBRYJ59N6cJocb_7Z:-LetSuOl8Pkc26Jb5-dQ:b9q9iyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1056)):*
<details><summary><i>Previously, ridwanmsharif (Ridwan Sharif) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I know what you mean but when you read this comment without context it can be confusing. I would just remove this line - it's up to the caller to make sure doing this makes sense, we just need to describe this function.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1053 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LetSfXL80s4dWjirAp-:-LetSfXL80s4dWjirAp0:b3lab9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/14a335ce4961f3a95dfc0184327a7dc109fe59b0/pkg/sql/opt/norm/custom_funcs.go#L1053)):*
> ```Go
> }
> 
> // mapSetOpFilter maps filters expressions to use the output columns of the
> ```

[nit] to replace occurrences of columns in `src` with the corresponding columns in `dst` (the two lists must have the same length).


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37418)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @justinj, and @rytaft)

___
*[pkg/sql/opt/norm/custom_funcs.go, line 1056 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LerBRYJ59N6cJocb_7Z:-Lew-__7-LQleK6ErsyP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aefce92b0055a2f7d6b7f077f5d1d080945e00f1/pkg/sql/opt/norm/custom_funcs.go#L1056)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I know what you mean but when you read this comment without context it can be confusing. I would just remove this line - it's up to the caller to make sure doing this makes sense, we just need to describe this function.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/custom_funcs.go, line 1053 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/37418#-LetSfXL80s4dWjirAp-:-Lew-bKB9RkiF-R0kPEF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14a335ce4961f3a95dfc0184327a7dc109fe59b0/pkg/sql/opt/norm/custom_funcs.go#L1053)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] to replace occurrences of columns in `src` with the corresponding columns in `dst` (the two lists must have the same length).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ridwanmsharif commented:
bors r+
